cpp/src/join/join_kernels.cuh

```diff
@@ -0,0 +1,433 @@
+ * @tparam multimap_type The type of the hash table
+ * @tparam output_index_type The datatype used for the indices in the output arrays
+ * @tparam block_size The number of threads per block for this kernel
+ * @tparam output_cache_size The side of the shared memory buffer to cache join output results
+ *
+ */
+/* ----------------------------------------------------------------------------*/
+template< join_type join_t,
+          typename multimap_type,
+          typename key_type,
+          typename output_index_type,
+          cudf::size_type block_size,
+          cudf::size_type output_cache_size>
+__global__ void probe_hash_table( multimap_type multi_map,
+                                  table_device_view build_table,
+                                  table_device_view probe_table,
+                                  row_hash hash_probe,
+                                  row_equality check_row_equality,
+                                  const cudf::size_type probe_table_num_rows,
+                                  output_index_type * join_output_l,
+                                  output_index_type * join_output_r,
+                                  numeric_scalar_device_view<cudf::size_type> current_idx,
+                                  const cudf::size_type max_size,
+                                  bool flip_results,
+                                  const output_index_type offset = 0)
+{
+  constexpr int num_warps = block_size/experimental::detail::warp_size;
+  __shared__ cudf::size_type current_idx_shared[num_warps];
+  __shared__ output_index_type join_shared_l[num_warps][output_cache_size];
+  __shared__ output_index_type join_shared_r[num_warps][output_cache_size];
+  output_index_type *output_l = join_output_l, *output_r = join_output_r;
+
+  if (flip_results) {
+      output_l = join_output_r;
+      output_r = join_output_l;
+  }
+
+  const int warp_id = threadIdx.x/experimental::detail::warp_size;
+  const int lane_id = threadIdx.x%experimental::detail::warp_size;
+
+  if ( 0 == lane_id )
+  {
+    current_idx_shared[warp_id] = 0;
+  }
+
+#if defined(CUDA_VERSION) && CUDA_VERSION >= 9000
+  __syncwarp();
+#endif
+
+  output_index_type probe_row_index = threadIdx.x + blockIdx.x * blockDim.x;
+
+#if defined(CUDA_VERSION) && CUDA_VERSION >= 9000
+  const unsigned int activemask = __ballot_sync(0xffffffff, probe_row_index < probe_table_num_rows);
+#endif
+  if ( probe_row_index < probe_table_num_rows ) {
+
+    const auto unused_key = multi_map.get_unused_key();
+    const auto end = multi_map.end();
+    auto found = end;
+
+    // Search the hash map for the hash value of the probe row using the row's
+    // hash value to determine the location where to search for the row in the hash map
+
+    // Only probe the hash table if the probe row is valid
+    hash_value_type probe_row_hash_value{0};
+    // Search the hash map for the hash value of the probe row
+    probe_row_hash_value = hash_probe(probe_row_index);
```

Comment:
I'm realizing that a future optimization should be to eliminate the redundant computation of the row hash values between the probe and join size estimation kernels. We can just compute this once and reuse it in both kernels.